storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,484 @@
+}
+
+// NewPruningStorer will return a new instance of PruningStorer without sharded directories' naming scheme
+func NewPruningStorer(args *PruningStorerArgs) (*PruningStorer, error) {
+	return initPruningStorer(args, "")
+}
+
+// NewShardedPruningStorer will return a new instance of PruningStorer with sharded directories' naming scheme
+func NewShardedPruningStorer(
+	args *PruningStorerArgs,
+	shardId uint32,
+) (*PruningStorer, error) {
+	shardIdStr := fmt.Sprintf("%d", shardId)
+	return initPruningStorer(args, shardIdStr)
+}
+
+// initPruningStorer will create a PruningStorer with or without sharded directories' naming scheme
+func initPruningStorer(
+	args *PruningStorerArgs,
+	shardIdStr string,
+) (*PruningStorer, error) {
+	var cache storage.Cacher
+	var db storage.Persister
+	var bf storage.BloomFilter
+	var err error
+
+	defer func() {
+		if err != nil && db != nil {
+			_ = db.Destroy()
+		}
+	}()
+
+	if args.NumOfActivePersisters < 1 {
+		return nil, storage.ErrInvalidNumberOfPersisters
+	}
+	if check.IfNil(args.Notifier) {
+		return nil, storage.ErrNilEpochStartNotifier
+	}
+	if check.IfNil(args.PersisterFactory) {
+		return nil, storage.ErrNilPersisterFactory
+	}
+
+	cache, err = storageUnit.NewCache(args.CacheConf.Type, args.CacheConf.Size, args.CacheConf.Shards)
+	if err != nil {
+		return nil, err
+	}
+
+	filePath := args.DbPath
+	if len(shardIdStr) > 0 {
+		filePath = filePath + shardIdStr
+	}
+	db, err = args.PersisterFactory.Create(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	var persisters []*persisterData
+	persisters = append(persisters, &persisterData{
+		persister: db,
+		path:      filePath,
+		isClosed:  false,
+	})
+
+	persistersMapByEpoch := make(map[uint32]*persisterData)
+	// TODO: get the starting epoch as a parameter
+	persistersMapByEpoch[0] = persisters[0]
+	pdb := &PruningStorer{
+		identifier:            args.Identifier,
+		fullArchive:           args.FullArchive,
+		activePersisters:      persisters,
+		persisterFactory:      args.PersisterFactory,
+		persistersMapByEpoch:  persistersMapByEpoch,
+		cacher:                cache,
+		bloomFilter:           nil,
+		dbPath:                filePath,
+		numOfEpochsToKeep:     args.NumOfEpochsToKeep,
+		numOfActivePersisters: args.NumOfActivePersisters,
+	}
+
+	if args.BloomFilterConf.Size != 0 { // if size is 0, that means an empty config was used so bloom filter will be nil
+		bf, err = storageUnit.NewBloomFilter(args.BloomFilterConf)
+		if err != nil {
+			return nil, err
+		}
+
+		pdb.bloomFilter = bf
+	}
+
+	err = pdb.activePersisters[0].persister.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	pdb.registerHandler(args.Notifier)
+
+	return pdb, nil
+}
+
+// Put adds data to both cache and persistence medium and updates the bloom filter
+func (ps *PruningStorer) Put(key, data []byte) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	ps.cacher.Put(key, data)
+
+	err := ps.activePersisters[0].persister.Put(key, data)
```

Comment:
shouldn't you add according to epoch ?

Reply:
Would it make sense to add data to a previous epoch ? for `get` and `has` it definitely makes sense, but is there any use case you would need to put in a certain epoch ?